### PR TITLE
Add DR for unstable plugin API

### PR DESCRIPTION
**Pull request by BrynCooke**
Adds a decision record for adding an unstable plugin API.
Aim to ratify by 2023-10-30

### CODE REVIEW

**Review by Geal — APPROVED** [review]

**Review by o0Ignition0o — APPROVED** [review]
This looks great!

Are there implications re Rhai / Coprocessors when these APIs graduate to unstable / public?

**Comment by BrynCooke** [comment]
> This looks great!
> 
> Are there implications re Rhai / Coprocessors when these APIs graduate to unstable / public?

Let's add something to the DR that explains graduation criteria.

Gut feeling is that to move from unstable to public we have to make sure it integrates with everything, but beyond that we can accept piecemeal integration.

File: dev-docs/decision-records/unstable-plugins.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+* pub (crate) PluginPrivate
+
+Demo here: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b4022dfeaeb920bc7ae94a94b7b9ea23
+
+This doesn't give us permission to not think about APIs we introduce, but it does enable us to unblock ourselves and our users in a safe way.
+
+## Acceptance criteria
+* Existing plugins should compile without modification.
+* Plugins external to the Router codebase cannot use the private API.
+* Plugins external to the router codebase can use the unstable API.
```
Comment (bnjjj) [inline dev-docs/decision-records/unstable-plugins.md:35]: Should we gate it under a feature called `router-unstable` like what we can find in `tracing` or `tokio` ?

File: dev-docs/decision-records/unstable-plugins.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+The order of introduction of new hook points would be:
+* Private - Does this work and make the router code cleaner?
+* Unstable - Available for use by plugins, but not guaranteed to be stable.
+* Public - Usable by all, and won't change.
+
+We will promote hook points from private to unstable to public as we gain confidence in them.
+
+## Hook points that could be introduced
+
+1. Query planning - Allows plugins tom implement caching, or other optimizations.
```
Comment (garypen) [inline dev-docs/decision-records/unstable-plugins.md:15]:
```suggestion
1. Query planning - Allows plugins to implement caching, or other optimizations.
```

**Review by garypen — CHANGES_REQUESTED** [review]

**Comment by garypen** [comment]
> This looks great!
> 
> Are there implications re Rhai / Coprocessors when these APIs graduate to unstable / public?

I think there would be. We should perhaps add to the DR that before graduating (to unstable? or to public?) from private we should have considered and decided what details (if any) would be accessible from coprocessors and Rhai. I think we'd also need to have good reasons to not expose such details via these mechanisms.

File: dev-docs/decision-records/unstable-plugins.md (hunk 1 above)
Comment (BrynCooke) [reply dev-docs/decision-records/unstable-plugins.md:35]: Don't think we'll feature flag it as much of the benefit is that we get feedback from not only users but ourselves. Our plugins would not be able to use the unstable APIs if we feature gate it.

**Comment by BrynCooke** [comment]
I added some discussion for graduation that can be discussed.

**Review by garypen — APPROVED** [review]

**Review by bnjjj — APPROVED** [review]

**Pull request merged**